docs_src/design/adr/0019-Service-List.md

```diff
@@ -0,0 +1,87 @@
+
+4. While Consul is often used as the configuration/registry service, EdgeX was constructed to operate off of local configuration in the absence of the central config/registry service (for development purposes or in order to reduce the resource needs of the platform).  The service list need should be provided when a central configuration/registry service like Consul is provided and even when services operate off local configuration.  Per #1 above, all services should get the service list information in the same way, but that way may change depending on the existence of a central configuration/registration service.
+
+### Considerations
+
+- In some cases, the list of services is needed before the services are up (security needs to know so appropriate tokens can be generated and handed out when the services bootstrap).
+- In some cases, a service may be “disabled” at runtime because of resource constraints (example: notification service is turned off as it is not used and there is a need to operate with limited memory and CPU capability).  Security would still need to know about disabled services in order to create a token for such services that may be enabled later.  However, the other services (or 3rd party applications) may need to know that they are disabled vs enabled.
+- In the case that the services are distributed (especially when device services run on different hosts - as is likely), access to a distributed file system or volume mount becomes more of an issue and requires additional technology.
+- The service list would have to be a list of services and service locations (URI).  Therefore, the service list will really be a map of service names (keys) and the location of the services URIs (values).  This will help support things like the API gateway setup as well.  In some cases, just a list of the service names is needed (in security token setup for example).  In this latter case, the service would just pick off the service keys.
+- Generation of security tokens for the new services under most of the conceived alternatives is complex.  In the current security bootstrapping, secret store setup will require a restart of all the services in order to add or remove a service.  The ORRA project and some adopters are clear that they would like to have the ability to add/remove services dynamically at runtime.
+- From a security perspective, being able to add/remove new services dynamically opens up a rather large attack vector and vulnerability.
+
+### Implementation Alternative Discussion
+
+- A service list could be made available in a configuration file in the file system (non-Docker) or volume mount (in a Docker situation).
+  - All services would get the configuration in a consistent way (#1)
+  - It would also mean that there is a single, authoritative list for all services and it could be made to be ACID safe (#2 and #3).
```

Comment:
> It would also mean that there is a single, authoritative list for all services and it could be made to be ACID safe (#2 and #3).
I disagree that concurrent updates to a configuration file could be made ACID compliant.

Reply:
Point taken.  There are technologies such as Transactional NTFS from Microsoft that make some claims with regard to transactional activity on files, but finding an easy to implement, ACID transactional, file based solution to the service list issue would be a challenge.  I'll note this.